astroquery/mast/cloud.py

```diff
@@ -0,0 +1,220 @@
+        """
+        Takes an `~astropy.table.Table` of data products and returns the associated cloud data uris.
+
+        Parameters
+        ----------
+        data_products : `~astropy.table.Table`
+            Table containing products to be converted into cloud data uris.
+        include_bucket : bool
+            Default True. When false returns the path of the file relative to the
+            top level cloud storage location.
+            Must be set to False when using the full_url argument.
+        full_url : bool
+            Default False. Return an HTTP fetchable url instead of a cloud uri.
+            Must set include_bucket to False to use this option.
+
+        Returns
+        -------
+        response : list
+            List of URIs generated from the data products, list way contain entries that are None
+            if data_products includes products not found in the cloud.
+        """
+
+        return [self.get_cloud_uri(product, include_bucket, full_url) for product in data_products]
+
+    def download_file(self, data_product, local_path, cache=True):
+        """
+        Takes a data product in the form of an  `~astropy.table.Row` and downloads it from the cloud into
+        the given directory.
+
+        Parameters
+        ----------
+        data_product :  `~astropy.table.Row`
+            Product to download.
+        local_path : str
+            The local filename to which toe downloaded file will be saved.
+        cache : bool
+            Default is True. If file is found on disc it will not be downloaded again.
+        """
+
+        s3 = self.boto3.resource('s3')
+        s3_client = self.boto3.client('s3')
+        bkt = s3.Bucket(self.pubdata_bucket)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            bucket_path = self.get_cloud_uri(data_product, False)
+        if not bucket_path:
+            raise Exception("Unable to locate file {}.".format(data_product['productFilename']))
+
+        # Ask the webserver (in this case S3) what the expected content length is and use that.
+        info_lookup = s3_client.head_object(Bucket=self.pubdata_bucket, Key=bucket_path, RequestPayer='requester')
+        length = info_lookup["ContentLength"]
+
+        if cache and os.path.exists(local_path):
+            if length is not None:
+                statinfo = os.stat(local_path)
+                if statinfo.st_size != length:
+                    log.warning("Found cached file {0} with size {1} that is "
+                                "different from expected size {2}"
+                                .format(local_path,
+                                        statinfo.st_size,
+                                        length))
+                else:
+                    log.info("Found cached file {0} with expected size {1}."
+                             .format(local_path, statinfo.st_size))
+                    return
```

Comment:
not needed

Reply:
This return statement is definitely needed, if it wasn't there function execution would continue to the `with ProgressBarOrSpinner` block and re-download the file even if a cached copy was found.

Reply:
oh, I totally missed... the rest of this function, I got blinded and thought the next section was a new function.  My bad.